zap/src/main/java/org/zaproxy/zap/extension/ascan/filters/impl/AbstractGenericScanFilter.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2019 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascan.filters.impl;
+
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiPredicate;
+import org.parosproxy.paros.Constant;
+import org.zaproxy.zap.extension.ascan.filters.FilterCriteria;
+import org.zaproxy.zap.extension.ascan.filters.FilterResult;
+import org.zaproxy.zap.extension.ascan.filters.ScanFilter;
+
+/**
+ * Abstract ScanFilter for handling generic filter usecases.
+ *
+ * @author KSASAN preetkaran20@gmail.com
+ * @since 2.9.0
+ */
+public abstract class AbstractGenericScanFilter<T> implements ScanFilter {
+
+    public static final String INCLUDE_FILTER_CRITERIA_MESSAGE_KEY =
+            "scan.filter.filtercriteria.include";
+    public static final String EXCLUDE_FILTER_CRITERIA_MESSAGE_KEY =
+            "scan.filter.filtercriteria.exclude";
+
+    private BiPredicate<Collection<T>, T> matcher;
```

Comment:
Sorry, didn't notice in the previous review, the original suggestion was to parameterize the filter data and the value (e.g. `BiPredicate<Collection<T>, V>`) to allow the filter data to be of a different type than the values being checked (as is the case of URL which uses `Pattern` and `String`).

Reply:
hmm correct. i was also thinking same, somehow missed. will do the change.